domain/consensus/processes/consensusstatemanager/pick_virtual_parents.go

```diff
@@ -34,7 +34,14 @@ func (csm *consensusStateManager) pickVirtualParents(stagingArea *model.StagingA
 	}
 	log.Debugf("The selected parent of the virtual is: %s", virtualSelectedParent)
 
-	candidates := candidatesHeap.ToSlice()
+	// Limit to maxBlockParents*3 candidates, that way we don't go over thousands of tips when the network isn't healthy.
+	// There's no specific reason for a factor of 3, and its not a consensus rule, just an estimation saying we probably
+	// don't want to consider and calculate 3 times the amount of candidates for the set of parents.
+	candidates := make([]*externalapi.DomainHash, 0, candidatesHeap.Len())
+	for len(candidates) < int(csm.maxBlockParents)*3 && candidatesHeap.Len() > 0 {
+		candidates = append(candidates, candidatesHeap.Pop())
+	}
```

Comment:
I think this could be slightly improved.
If candidatesHeap is smaller than maxBlockParents then `candidates := candidatesHeap.ToSlice()`
Otherwise, `candidates := make([]*externalapi.DomainHash, 0, maxBlockParents)...`